### ISSUE TITLE
Pyre doesn't seem to understand universal quantification

### ISSUE THREAD

**Issue by vthemelis** [facebook/pyre-check#771]
# Pyre Bug

**Bug description**
I'm trying to type a lambda-based encoding of `cons` from Lisp:

```Python3
from typing import Callable, TypeVar

T = TypeVar('T')
U = TypeVar('U')
V = TypeVar('V')

def cons(a: T, b: U) -> Callable[[Callable[[T, U], V]], V]:
    return lambda f : f(a, b)
```

This fails with:
```
Invalid type variable [34]: The type variable `Variable[V]` isn't present in the function's parameters.
```

`mypy` seems to have no issue type-checking this

**Reproduction steps**
Run pyre with the snippet above

**Expected behavior**
The above should type-check.

**Comment by kinto0** [comment]
Hi! As far as I know, TypeVars must be present in the function's parameters. To work around this, you could use a similar strategy as decorator factories from our [documentation for this error](https://pyre-check.org/docs/errors/#decorator-factories). I did it [here](https://pyre-check.org/play?input=%23%20Pyre%20is%20being%20run%20in%20gradual%20typing%20mode%3A%20https%3A%2F%2Fpyre-check.org%2Fdocs%2Ftypes-in-python%2F%23gradual-typing%0A%23%20Use%20the%20%60%23%20pyre-strict%60%20header%20to%20run%20in%20strict%20mode%2C%20which%20requires%20annotations.%0A%0Afrom%20typing%20import%20Callable%2C%20Protocol%2C%20TypeVar%0A%0AT%20%3D%20TypeVar(%27T%27)%0AU%20%3D%20TypeVar(%27U%27)%0AV%20%3D%20TypeVar(%27V%27)%0A%0Aclass%20MyCallableProtocol(Protocol)%3A%0A%20%20%20%20def%20__call__(self%2C%20f%3A%20Callable%5B%5BT%2C%20U%5D%2C%20V%5D)%20-%3E%20V%3A%20...%0A%0Adef%20cons(first%3A%20T%2C%20rest%3A%20U)%20-%3E%20MyCallableProtocol%3A%0A%20%20%20%20return%20lambda%20f%20%3A%20f(first%2C%20rest)%0A)

**Comment by vthemelis** [comment]
@kinto0 thank you ver much! Indeed, I should have seen this earlier.

With some modification, it looks like I can type check `car` and `cdr` as well:

Pyre playground (here)[https://pyre-check.org/play?input=%23%20Pyre%20is%20being%20run%20in%20gradual%20typing%20mode%3A%20https%3A%2F%2Fpyre-check.org%2Fdocs%2Ftypes-in-python%2F%23gradual-typing%0A%23%20Use%20the%20%60%23%20pyre-strict%60%20header%20to%20run%20in%20strict%20mode%2C%20which%20requires%20annotations.%0A%0Afrom%20typing%20import%20Callable%2C%20Generic%2C%20TypeVar%2C%20Protocol%0A%0AT%20%3D%20TypeVar(%27T%27)%0AU%20%3D%20TypeVar(%27U%27)%0AV%20%3D%20TypeVar(%27V%27)%0A%0Aclass%20MyCallableProtocol(Protocol%2C%20Generic%5BT%2C%20U%5D)%3A%0A%20%20%20%20def%20__call__(self%2C%20f%3A%20Callable%5B%5BT%2C%20U%5D%2C%20V%5D)%20-%3E%20V%3A%20...%0A%0Adef%20cons(a%3A%20T%2C%20b%3A%20U)%20-%3E%20MyCallableProtocol%5BT%2C%20U%5D%3A%0A%20%20%20%20return%20lambda%20f%20%3A%20f(a%2C%20b)%0A%0Adef%20car(cons%3A%20MyCallableProtocol%5BT%2C%20U%5D)%20-%3E%20T%3A%0A%20%20%20%20cont%3A%20Callable%5B%5BT%2C%20U%5D%2C%20T%5D%20%3D%20lambda%20x%2C%20_%3A%20x%0A%20%20%20%20return%20cons(cont)%0A%0Adef%20cdr(cons%3A%20MyCallableProtocol%5BT%2C%20U%5D)%20-%3E%20U%3A%0A%20%20%20%20cont%3A%20Callable%5B%5BT%2C%20U%5D%2C%20U%5D%20%3D%20lambda%20_%2C%20y%3A%20y%0A%20%20%20%20return%20cons(cont)%0A%0A]

Something that may be of interest is that I need to write `car` like:

```
def car(cons: MyCallableProtocol[T, U]) -> T:
    cont: Callable[[T, U], T] = lambda x, _: x
    return cons(cont)
```
rather than

```
def car(cons: MyCallableProtocol[T, U]) -> T:
    return cons(lambda x, _: x)
```

to get full type safety. Without this change, changing the return type of `car` from `T` to something else doesn't elicit a type error. 

In other words, it doesn't seem like pyre can infer the type of the lambda correctly.


Thanks a lot for the help!
